src/main/java/com/aws/greengrass/mqttclient/MqttClient.java

```diff
@@ -442,37 +466,72 @@ public synchronized void unsubscribe(UnsubscribeRequest request)
             return future;
         }
 
+        SpoolMessage spooledMessage;
         try {
-            spool.addMessage(request);
+            spooledMessage = spool.addMessage(request);
             triggerSpooler();
         } catch (InterruptedException | SpoolerStoreException e) {
             logger.atDebug().log("Fail to add publish request to spooler queue", e);
             future.completeExceptionally(e);
             return future;
         }
-        return CompletableFuture.completedFuture(0);
+        return spooledMessage.getFuture();
+    }
+
+    protected boolean isValidPublishRequest(PublishRequest request) throws SpoolerStoreException {
+        // Payload size should be smaller than MQTT maximum message size
+        int messageSize = request.getPayload().length;
+        if (messageSize > MAXIMUM_SIZE_OF_MQTT_MESSAGE_SIZE) {
+            throw new SpoolerStoreException(String.format("The message size exceeds the maximum size of %d ",
+                    MAXIMUM_SIZE_OF_MQTT_MESSAGE_SIZE));
+        }
+
+        // Topic should not contain wildcard characters
+        String topic = request.getTopic();
+        if (topic.contains("#") || topic.contains("+")) {
+            throw new SpoolerStoreException("The topic of publish request should not contain wildcard characters.");
+        }
+
+        // Topic should not have no more than maximum number of forward slashes (/)
+        if (topic.chars().filter(num -> num == '/').count() > MAXIMUM_NUMBER_OF_FORWARD_SLASHES) {
```

Comment:
```
A topic in a publish or subscribe request can have no more than 7 forward slashes (/). This excludes the first 3 slashes in the mandatory segments for Basic Ingest topics ($AWS/rules/rule-name/).
```

We probably want to set our maximum to 10 in that case, or check for the reserved topic.